### PR TITLE
Allow AOT for more Intel GPUs using SYCL

### LINKED ISSUE #4317: General Architecture Handling for Kokkos
This issue references discussions in PR #4285.  The larger design question is how to deal with the emerging diversity of architectures.  We need a future-proof design.  A unified solution will work for:

**CPU**

- Many host architectures supported:  https://github.com/kokkos/kokkos/blob/master/cmake/kokkos_arch.cmake#L32
- See list of supported CPU 
- Backends: OpenMP

**GPU**

- INTEL
  - Backends:  SYCL, OpenMPTarget
  - Platforms:  Aurora  
- AMD
  - Backends: HIP, OpenMPTarget
  - Platforms: Frontier, El Capitan
- NVIDIA
  - Backends: Cuda, OpenMPTarget
  - Platforms:  Sierra, Perlmutter (LBL NERSC)

### PR DESCRIPTION
**Pull request by masterleinad**
This pull request extends the lists of Intel GPUs available to the ones in https://software.intel.com/content/www/us/en/develop/documentation/oneapi-dpcpp-cpp-compiler-dev-guide-and-reference/top/compilation/ahead-of-time-compilation.html.
The first commit uses a list approach as we usually do. Note that `skl` is effectively replaced by `gen9`. Thus, we are compiling for more configurations now. 
Since this approach appeared tedious and restrictive, the second commit just allows the user to provide the string used when invoking `ocloc`.

### CODE REVIEW

**Review by nliber — APPROVED** [review]

**Comment by masterleinad** [comment]
@nliber What does your approval mean? Which version are you in favor of?

File: cmake/kokkos_arch.cmake [hunk 1]
```diff
@@ -67,9 +67,16 @@ KOKKOS_ARCH_OPTION(ZEN3            HOST "AMD Zen3 architecture")
...
-
+KOKKOS_ARCH_OPTION(INTEL_GEN       GPU  "Intel GPUs Gen9+ (deprecated)")
 
+KOKKOS_OPTION(ARCH_INTEL_GPU "" STRING "Intel GPU architecture used for ahead-of-time compilation")
+IF(KOKKOS_ARCH_INTEL_GEN)
+  # deprecated
+  SET(KOKKOS_ARCH_INTEL_GPU skl CACHE STRING "Intel GPU architecture used for ahead-of-time compilation" FORCE)
+ENDIF()
+IF(KOKKOS_ARCH_INTEL_GPU)
+  LIST(APPEND KOKKOS_ENABLED_ARCH_LIST "INTEL_GPU_${KOKKOS_ARCH_INTEL_GPU}")
```
Comment (crtrott) [inline cmake/kokkos_arch.cmake:78]: should this be upper cased?

**Comment by crtrott** [comment]
I think this is a bigger discussion we should have. It might be worthwhile for us to discuss switching completely to a different scheme. I.e. we do have a KOKKOS_ARCH define with a value, and maybe a KOKKOS_ARCH_NVIDIA_GPU, KOKKOS_ARCH_INTEL_GPU, KOKKOS_ARCH_AMD_GPU. On the other hand if we ever wanted to switch to a different model where we allow multi arch generation that could be problematic too. Also should we then have KOKKOS_ARCH_INTEL_CPU and KOKKOS_ARCH_AMD_CPU? What does it mean to compile for generic AVX which works on both Intel and AMD CPUs?

**Comment by masterleinad** [comment]
I am open to whatever we think is best but I am slightly opposed to adding >10 architectures for Intel GPUs. The approach taken here already allows multi-arch generation for Intel GPUs:
```
-device 0xFF20     ; will compile 1 target (tgllp)
-device skl        ; will compile 1 target
-device skl,icllp  ; will compile 2 targets
-device skl-icllp  ; will compile all targets in range (inclusive)
-device skl-       ; will compile all targets newer/same as provided
-device -skl       ; will compile all targets older/same as provided
-device gen9       ; will compile all targets matching the same gen
-device gen9-gen11 ; will compile all targets in range (inclusive)
-device gen9-      ; will compile all targets newer/same as provided
-device -gen9      ; will compile all targets older/same as provided
-device *          ; will compile all targets known to ocloc
```
and I could imagine us using a similar approach (comma-separated values) for other architecture types.

**Comment by dalg24** [comment]
For those who like me wondered what AOT means: Ahead of Time Compilation

File: cmake/kokkos_arch.cmake [hunk 2]
```diff
@@ -67,9 +67,16 @@ KOKKOS_ARCH_OPTION(ZEN3            HOST "AMD Zen3 architecture")
 KOKKOS_ARCH_OPTION(VEGA900         GPU  "AMD GPU MI25 GFX900")
 KOKKOS_ARCH_OPTION(VEGA906         GPU  "AMD GPU MI50/MI60 GFX906")
 KOKKOS_ARCH_OPTION(VEGA908         GPU  "AMD GPU MI100 GFX908")
-KOKKOS_ARCH_OPTION(INTEL_GEN       GPU  "Intel GPUs Gen9+")
-
+KOKKOS_ARCH_OPTION(INTEL_GEN       GPU  "Intel GPUs Gen9+ (deprecated)")
 
+KOKKOS_OPTION(ARCH_INTEL_GPU "" STRING "Intel GPU architecture used for ahead-of-time compilation")
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:72]: Why is it a `KOKKOS_OPTION` and not a `KOKKOS_ARCH_OPTION`?

File: cmake/kokkos_arch.cmake [hunk 3]
```diff
@@ -67,9 +67,16 @@ KOKKOS_ARCH_OPTION(ZEN3            HOST "AMD Zen3 architecture")
...
+IF(KOKKOS_ARCH_INTEL_GEN)
+  # deprecated
+  SET(KOKKOS_ARCH_INTEL_GPU skl CACHE STRING "Intel GPU architecture used for ahead-of-time compilation" FORCE)
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:75]: Should we emit a warning?

**Review by dalg24 — COMMENTED** [review]
Would you not want to validate the string in the 2nd approach?

File: cmake/kokkos_arch.cmake (hunk 2 above)
Comment (masterleinad) [reply cmake/kokkos_arch.cmake:72]: I explicitly want a `STRING` here, not a `BOOL`.

File: cmake/kokkos_arch.cmake (hunk 3 above)
Comment (masterleinad) [reply cmake/kokkos_arch.cmake:75]: Possibly, I'll wait for the outcome of the discussion before continuing here, though.

**Comment by masterleinad** [comment]
At least for `SYCL` multi-arch builds are not a problem at the moment (and not in the foreseeable future) if we at least use the same device type. There are some optimizations for Intel GPUs that we can't do for `Nvidia` GPUs, though.

That being said, the point here is not that much allowing multi-arch builds but rather a more flexible approach avoiding introducing a large number of `CMake` architecture options. At least for `SYCL` I fail to see a downside of this approach (other than being inconsistent with the way we are handling architectures otherwise and requiring the user to provide some obscure but well documented strings).

**Comment by ajpowelsnl** [comment]
Hi @masterleinad - Would you like to discuss this further in our Wed. meeting?  If so, would it be OK if I made you the champion of this issue?

**Comment by masterleinad** [comment]
> Hi @masterleinad - Would you like to discuss this further in our Wed. meeting? If so, would it be OK if I made you the champion of this issue?

AFAIK, you were supposed to create an issue for discussing the way to specify architectures all backends more generally. I want to at least come to a conclusion with respect to SYCL soon and am happy to explain it some more.

**Comment by crtrott** [comment]
Do we want to allow the SYCL backend to have a different style of defining architectures than other backends (i.e. what this PR proposes)?
SF F N A SA
1 0 6 5 0

**Comment by masterleinad** [comment]
I reworked the pull request to introduce `CMake` options for all the Intel GPU architectures I want to add here.

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
This and #4305 need to be synchronized whenever we merge one of them.

**Review by crtrott — APPROVED** [review]

**Pull request merged**